### ISSUE TITLE
Selecting from sub query is returning unexpected data in specific cases

### ISSUE THREAD

**Issue by dannymsullivan** [influxdata/influxdb#7874]
### Bug report

System info: Influxdb 1.2, Max OS 10.12.2

Steps to reproduce:

1. Seed influx with [data](https://gist.github.com/dsully360/de3d33fa67bdf2e9679abb81d19a85e5)
2. Run ```select * from workflows group by work_id order by time desc limit 1```. This is the sub query, which returns the correct data I want to filter on.
3. Then run the full query ```SELECT * FROM (select * from workflows group by work_id order by time desc limit 1 ) WHERE state = 'error'```. This returns expected data.
4. Then run ```SELECT * FROM (select * from workflows group by work_id order by time desc limit 1 ) WHERE state = 'render'```. This returns expected data.
5. Then run ```SELECT * FROM (select * from workflows group by work_id order by time desc limit 1 ) WHERE state = 'send'``` This returns unexpected data.

__Expected behavior:__ I expect the sub query to only return the items that match the where clause, after the sub query sorts them and limits them to one.

__Actual behavior:__ It returns data that should not be in the subquery.

__Additional info:__ It seems like the sub query is not carrying over the correct data possibly.
![valid-1](https://cloud.githubusercontent.com/assets/2847232/22266380/c47326de-e24d-11e6-99ab-1d759088b1ea.png)
![valid-2](https://cloud.githubusercontent.com/assets/2847232/22266383/c6cd076a-e24d-11e6-8acd-274e3c462fc4.png)
![bad-data](https://cloud.githubusercontent.com/assets/2847232/22266384/c88c4930-e24d-11e6-8bac-979032f4d65e.png)
![screen shot 2017-01-24 at 4 12 34 pm](https://cloud.githubusercontent.com/assets/2847232/22267001/184d59a8-e250-11e6-9b71-b486bc8d43cb.png)

**Comment by jsternberg** [comment]
Can you try `select * from workflows group by work_id order by time desc limit 1` on its own? I have a feeling there will be no entries returned that have `state = 'send'`, but I can bet there will be ones that say `state = 'render'`.

You might have to put the `WHERE state = 'send'` in the inner query instead of the outer query.

**Comment by dsully360** [comment]
Ok, lemme test it out, will report back.

**Comment by dsully360** [comment]
```select * from workflows group by work_id order by time desc limit 1``` returns what I expect. The issue I am having is, I need to get the most recent state, by time, for each `work_id` first. Then when I have that data, I can select only items that match the where clause, and return those. If I put the where clause in the sub query, then group it and limit it, I am no longer only matching on the most recent state of the work items, but any item that matches.

**Comment by dsully360** [comment]
![screen shot 2017-01-25 at 12 50 18 pm](https://cloud.githubusercontent.com/assets/2847232/22302549/41e08e54-e2fd-11e6-98f3-d9073ddb4641.png)
![screen shot 2017-01-25 at 12 50 05 pm](https://cloud.githubusercontent.com/assets/2847232/22302552/4369dd02-e2fd-11e6-95fe-d6889de1efe4.png)

**Comment by dsully360** [comment]
I think this https://github.com/influxdata/influxdb/issues/7875, #2, is referring to the same issue. Limit is possibly getting ignored when Group By is in the subquery.

**Comment by jsternberg** [comment]
Is there any chance you can show me the results of just the inner query and then the result with it being used as a subquery? The queries you are using are split between different images so it's not easy for me to identify which ones relate to each other. So if you have this as your query:

```
SELECT * FROM (select * from workflows group by work_id order by time desc limit 1 ) WHERE state = 'send'
```

What is the result of `select * from workflows group by work_id order by time desc limit 1` and then what is the result of the entire query?

**Comment by dsully360** [comment]
Got ya, coming right up.

**Comment by dsully360** [comment]
![screen shot 2017-01-25 at 3 32 56 pm](https://cloud.githubusercontent.com/assets/2847232/22308101/92e8a118-e313-11e6-8e77-b5e21c642cee.png)

**Comment by jsternberg** [comment]
I see the wildcard expansion seems to be a bit weird, but what's the mistake in the data? It looks like the final query only contains data where `state = 'send'`. What output do you expect?

**Comment by dsully360** [comment]
The subquery doesn't have any data with `send`. I thought the main query was selecting on that data, which shouldn't match anything.

**Comment by dsully360** [comment]
It looks to me like the main query is just selecting on all the data and ignoring the subquery data.

**Comment by jsternberg** [comment]
Oh, I see now. I read the original picture wrong and saw `send` in another column. I'll look at it. You're right, this looks like a bug.

**Comment by dsully360** [comment]
Ok sounds good 👍

**Comment by jsternberg** [comment]
I think this might be the same as #7885. The `last()` function internally changes itself to a `order by time desc limit 1` query for speed and I think they're the same issue.

**Comment by dsully360** [comment]
Ok great, thanks. I was looking at that issue yesterday, looked similar in nature.

**Comment by dsully360** [comment]
Are you aware of another way to accomplish what we are trying to do here? I am open to suggestions, if this fix might be a few months off.

**Comment by jsternberg** [comment]
I don't think so. I just came up with a solution for the `LIMIT` and `OFFSET` problem. If you need `SLIMIT` and `SOFFSET`, you might have to wait a bit while I figure out how to structure that in the code.

The current solution isn't what I would call ideal (I don't think it will perform nearly as well as I would like, but I don't have the benchmarks), but it should work for being able to get the data. I just created the PR a couple of minutes ago in #7892 so please try it out.

**Comment by dsully360** [comment]
Ok awesome, thanks for cranking that out.

**Comment by jsternberg** [comment]
I'm trying this out and while the other issue is fixed, this one isn't because this specific type of query doesn't properly evaluate the outer where clause. I'm looking into it.

**Comment by dsully360** [comment]
Ok got ya, thanks.

**Comment by dsully360** [comment]
Any word on this?

**Comment by jsternberg** [comment]
I'm still working on it. The conditions are currently a bit screwy and require some extra work.

**Comment by dsully360** [comment]
Ok no worries, thanks for the update.

**Comment by dsully360** [comment]
Hey Jonathan, just checking in on this, do you think this is fixable in the near future? Even a few weeks to a couple months would work for me. I am just trying to decide if I can move ahead with Influxdb, but this functionality is  key. Any thoughts are greatly appreciated.

**Comment by dsully360** [comment]
Looks like this is fixed, this ```select * from (select * from workflows where state = 'error' group by work_id order by time desc limit 1)``` works, pretty sure.

**Comment by jsternberg** [comment]
I think this is now fixed and will be merged back to master soon. I'm going to close this. There may be remaining issues with how conditions are evaluated so please comment if this is still not working.

**Comment by dsully360** [comment]
I am using 1.2.1~rc1, and it doesn't seem to be working as expected. I know there are other release candidates out there, but just wanted to let you know.

<img width="1101" alt="screen shot 2017-02-14 at 10 48 47 am" src="https://cloud.githubusercontent.com/assets/2847232/22936438/876fb41c-f2a3-11e6-9767-abd41a5b3a40.png">

**Comment by jsternberg** [comment]
I'm going to reopen this since the conditions are screwy. This issue had two problems. One was the limit problem and one was a condition evaluation.

**Comment by dsully360** [comment]
Ok cool, thanks.

**Comment by dannymsullivan** [comment]
Just wanted to check in on this. I haven't pulled down the code in a while, any updates?

**Comment by jsternberg** [comment]
No progress on this at the moment. There's a PR that fixes the issue, but it also slows down the performance of some of the queries. I'll be trying to get this fix into 1.3.

**Comment by dannymsullivan** [comment]
Ok got ya, thanks.

**Comment by bashnesnos** [comment]
Hi @jsternberg , could this also be a root-cause for a query like this:

`SELECT (sum(ok_count) + sum(tol_count)/2)/sum(total_count)*100 as total 
FROM 
(SELECT count(total) as ok_count FROM session_load WHERE page = 'cart' AND time > now() - 30m AND total < 7000 GROUP BY time(1m),page fill(0)), 
(SELECT count(total) as tol_count FROM session_load WHERE page = 'cart' AND time > now() - 30m AND total >= 7000 AND total < 28000 GROUP BY time(1m),page fill(0)), 
(SELECT count(total) as total_count FROM session_load WHERE page = 'cart' AND time > now() - 30m GROUP BY time(1m),page fill(0)) 
WHERE time > now() - 30m group by time(1m),page fill(0)`

It works without grouping by a tag, but once I add the grouping by tag the math calculations are incorrect.
If I specify a concrete single tag - the results are correct.

**Comment by jsternberg** [comment]
This should be fixed by #8081.

**Comment by jsternberg** [comment]
Fixed by #8081.
